### PR TITLE
feat: make `BLOCKHASH` return only last blockhash

### PR DESCRIPTION
**Pull request by NazariiDenha**
Change behavior of `BLOCKHASH` accordingly to https://www.notion.so/scrollzkp/Differences-between-Scroll-and-Ethereum-aaab25ed6b734e50a61c4cce27e33096#ab6e0d856f1a45ea85997c668bd1e244

### CODE REVIEW

**Comment by ChuhanJin** [comment]
Can one of the admins verify this patch?

File: core/vm/runtime/runtime_test.go [hunk 1]
```diff
@@ -274,7 +274,7 @@ func (d *dummyChain) GetHeader(h common.Hash, n uint64) *types.Header {
 // requires access to a chain reader.
 func TestBlockhash(t *testing.T) {
 	// Current head
-	n := uint64(1000)
+	n := uint64(20)
```
Comment (Thegaram) [inline core/vm/runtime/runtime_test.go]: Let's not change this.

File: core/vm/runtime/runtime_test.go [hunk 2]
```diff
@@ -301,7 +301,7 @@ func TestBlockhash(t *testing.T) {
 				bytes32 zero;
 				zero = blockhash(x); // Should be zeroes
 				first = blockhash(x-1);
-				for(uint256 i = 2 ; i < 260; i++){
+				for(uint256 i = 2 ; i < 16; i++){
```
Comment (Thegaram) [inline core/vm/runtime/runtime_test.go]: If you want to change this, you'll need to recompile the contract (e.g. using Remix) and change `data` below.

Comment (Thegaram) [reply core/vm/runtime/runtime_test.go]: But it's fine to keep it as 260 (all those hashes should return 0).

File: core/vm/runtime/runtime_test.go [hunk 3]
```diff
@@ -334,13 +334,13 @@ func TestBlockhash(t *testing.T) {
...
+	if first.Uint64() != 19 {
+		t.Fatalf("second block should be 19, got %d (%x)", first, ret[32:64])
 	}
-	if last.Uint64() != 744 {
-		t.Fatalf("last block should be 744, got %d (%x)", last, ret[64:96])
+	if last.Uint64() != 0 {
+		t.Fatalf("last block should be 0, got %d (%x)", last, ret[64:96])
 	}
-	if exp, got := 255, chain.counter; exp != got {
+	if exp, got := 0, chain.counter; exp != got {
```
Comment (Thegaram) [inline core/vm/runtime/runtime_test.go:343]: What is `chain.counter`?

**Comment by Thegaram** [comment]
add to whitelist

File: core/vm/runtime/runtime_test.go (hunk 3 above)
Comment (NazariiDenha) [reply core/vm/runtime/runtime_test.go:343]: it is number of times when test queried chain info, e.g. called `GetHeader` method of chain

Comment (Thegaram) [reply core/vm/runtime/runtime_test.go:343]: Let's make sure we understand why this is 0, not 1.

**Review by Thegaram — DISMISSED** [review]

File: core/vm/runtime/runtime_test.go (hunk 3 above)
Comment (NazariiDenha) [reply core/vm/runtime/runtime_test.go:343]: https://github.com/scroll-tech/go-ethereum/blob/69c291cf7ac43be7e10979457e4137af0c63ce0e/core/evm.go#L84-L89

I think here is explanation, `GetHashFn` just caches latest `header` here

File: core/vm/instructions.go [hunk 4]
```diff
@@ -440,10 +440,10 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 	}
 	var upper, lower uint64
 	upper = interpreter.evm.Context.BlockNumber.Uint64()
-	if upper < 257 {
+	if upper < 1 {
```
Comment (icemelon) [inline core/vm/instructions.go]: This should be 2, right?

```suggestion
	if upper < 2 {
```

Comment (Thegaram) [reply core/vm/instructions.go]: Both versions result in `lower = 0`.

Comment (NazariiDenha) [reply core/vm/instructions.go]: Changed to `< 2`, because it's more consistent with previous logic

**Review by icemelon — DISMISSED** [review]

**Review by icemelon — APPROVED** [review]

**Review by 0xmountaintop — APPROVED** [review]

**Pull request merged**